### ISSUE TITLE
Expand cache services

### ISSUE THREAD

**Issue by kilchenmann** [dasch-swiss/knora-api-js-lib#106]
Can you expand the cache service with the following methods:

- get project by shortcode
- get project members by shortcode
- get project groups by shortcode
- get all users
- get user by username
- get ontology by iri

**Comment by tobiasschweizer** [comment]
- get ontology by iri -> use `OntologyCache.getOntology(ontologyIri: string)`
- get project by shortcode -> `ProjectsEndpoint.getProjectByShortname(shortname: string)` -> I can add a cache for this
- get project members by shortcode (`ProjectsEndpoint.getProjectMembersByShortcode(shortcode: string)`) -> I can add a cache for this
- get user by username -> I could adapt the `UserCache` which gets users by IRI
- get project groups by shortcode: Which endpoint and which method would have to be called?
- get all users (`UsersEndpoint.getUsers`the) -> if we call that method, we don't need a cache for users at all. Could you make an Angular service that just calls this method once?

**Comment by tobiasschweizer** [comment]
> get project groups by shortcode: Which endpoint and which method would have to be called?

@kilchenmann Does the endpoint already exist or does it have to be added?

**Comment by kilchenmann** [comment]
I'm sorry, didn't saw your previous answer. 

> get project groups by shortcode: Which endpoint and which method would have to be called?

Hm, it's a mistake from my side. there is now route or endpoint to get project specific groups. We have to use the `groupsEndpoint.getGroups` and filter by `project.id`. In this case the cache contains all groups. So, it should be a `getGroups cache`.

**Comment by tobiasschweizer** [comment]
> get all users (UsersEndpoint.getUsersthe) -> if we call that method, we don't need a cache for users at all. Could you make an Angular service that just calls this method once?

@kilchenmann Do we still need a user cache then?

**Comment by kilchenmann** [comment]
> get all users (UsersEndpoint.getUsersthe) -> if we call that method, we don't need a cache for users at all. Could you make an Angular service that just calls this method once?

Not sure. I have to run this request many times — not only to list all users. I need a list of all users also in the user-form to keep email and username unique. I compare new entries with the existing ones. But I can make my one cache for this.

**Comment by tobiasschweizer** [comment]
I think I could make a simple cache in the admin endpoint that requests all users when the endpoint is created and stores them all so you can ask for specific users without any further request to Knora. Of course, when user data changes or new users are added, the entries have to be updated.

I think this use case is simpler than the ones I have implemented so far so I will make a separate class for this.

**Comment by kilchenmann** [comment]
Keep in mind, that only system admin or project admin get an answer from usersEndpoint.getUsers.

**Comment by tobiasschweizer** [comment]
good point, so you need to log in. I will think of a good way to do this.

**Comment by kilchenmann** [comment]
Hm, but I don't think we'll need the list of all users outside of the admin interface. So, your solution should be fine.

**Comment by tobiasschweizer** [comment]
So we would have a `UserCache` by Iri so we could resolve `attachedToUser` props and use `getUsers()` only within the admin area?
